test/integration/cds_integration_test.cc

```diff
@@ -177,37 +134,233 @@ INSTANTIATE_TEST_SUITE_P(IpVersionsClientType, CdsIntegrationTest, GRPC_CLIENT_I
+      : CdsIntegrationTest(ipVersion(), ConfigHelper::discoveredClustersBootstrap("DELTA_GRPC")) {}
+
+  void TearDown() override {
+    cleanUpXdsConnection();
+    test_server_.reset();
+    fake_upstreams_.clear();
+  }
+
+  void giveInitialCluster() override {
+    EXPECT_TRUE(compareDeltaDiscoveryRequest(Config::TypeUrl::get().Cluster, {}, {}));
+    sendDeltaDiscoveryResponse<envoy::api::v2::Cluster>({cluster1_}, {}, "55");
+  }
+};
+
+INSTANTIATE_TEST_CASE_P(IpVersionsClientType, DeltaCdsIntegrationTest,
+                        GRPC_CLIENT_INTEGRATION_PARAMS);
+
+// 1) Envoy starts up with no static clusters (other than the CDS-over-gRPC server).
+// 2) Envoy is told of a cluster via CDS.
+// 3) We send Envoy a request, which we verify is properly proxied to and served by that cluster.
+// 4) Envoy is told that cluster is gone.
+// 5) We send Envoy a request, which should 503.
+// 6) Envoy is told that the cluster is back.
+// 7) We send Envoy a request, which we verify is properly proxied to and served by that cluster.
+TEST_P(DeltaCdsIntegrationTest, CdsClusterUpDownUp) {
+  // Calls CdsIntegrationTest::initialize(), which includes establishing a listener, route, and
+  // cluster.
+  testRouterHeaderOnlyRequestAndResponse(nullptr, UpstreamIndex1, "/cluster1");
+
+  // Tell Envoy that cluster_1 is gone.
+  EXPECT_TRUE(compareDeltaDiscoveryRequest(Config::TypeUrl::get().Cluster, {}, {}));
+  sendDeltaDiscoveryResponse<envoy::api::v2::Cluster>({}, {ClusterName1}, "42");
+  // We can continue the test once we're sure that Envoy's ClusterManager has made use of
+  // the DiscoveryResponse that says cluster_1 is gone.
+  test_server_->waitForCounterGe("cluster_manager.cluster_removed", 1);
+
+  // Now that cluster_1 is gone, the listener (with its routing to cluster_1) should 503.
+  BufferingStreamDecoderPtr response = IntegrationUtil::makeSingleRequest(
+      lookupPort("http"), "GET", "/cluster1", "", downstream_protocol_, version_, "foo.com");
+  ASSERT_TRUE(response->complete());
+  EXPECT_STREQ("503", response->headers().Status()->value().c_str());
+
+  cleanupUpstreamAndDownstream();
+  codec_client_->waitForDisconnect();
+
+  // Tell Envoy that cluster_1 is back.
+  EXPECT_TRUE(compareDeltaDiscoveryRequest(Config::TypeUrl::get().Cluster, {}, {}));
+  sendDeltaDiscoveryResponse<envoy::api::v2::Cluster>({cluster1_}, {}, "413");
+
+  // We can continue the test once we're sure that Envoy's ClusterManager has made use of
+  // the DiscoveryResponse describing cluster_1 that we sent. Again, 2 includes CDS server.
   test_server_->waitForGaugeGe("cluster_manager.active_clusters", 2);
 
   // Does *not* call our initialize().
-  testRouterHeaderOnlyRequestAndResponse(nullptr, UpstreamIndex);
+  testRouterHeaderOnlyRequestAndResponse(nullptr, UpstreamIndex1, "/cluster1");
+
+  cleanupUpstreamAndDownstream();
+}
+
+// Tests adding a cluster, adding another, then removing the first.
+TEST_P(DeltaCdsIntegrationTest, TwoClusters) {
```

Comment:
Why isn't something like `TwoClusters` a `TEST_P`, with the two different CDS implementations the parameters?

Reply:
Although almost all the code is the same, the differences are that analogous but different functions, even with different argument lists, are being called. (compareDiscoveryRequest,sendDiscoveryResponse, and Delta). If it's worth it to have the usage of the parameter just be `if (parameter) { one block of code } else { other block }`, then that would work. Although, then what would happen with VersionsRememberedAfterReconnect, which genuinely only makes sense for Delta? Just have it return/pass early when the parameter isn't delta?

Reply:
I guess the question is how many tests will end up being almost the same between delta and non-delta? If a lot, it could be worth it. This is what we did with https://github.com/envoyproxy/envoy/blob/master/test/common/config/subscription_impl_test.cc and the various xDS types. Your call really, just offering this as a potential maintainability improvement.